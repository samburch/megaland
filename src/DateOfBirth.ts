interface ValueObject {
  date: string;
  age?: number;
}

export class DateOfBirth implements ValueObject {
  age?: number;
  constructor(public date: string) {
    this.age = DateOfBirth.getAge(date);
    this.date = DateOfBirth.isValidDate(date);
  }

  static getAge(value: string): number {
    return new Date().getFullYear() - new Date(value).getFullYear();
  }

  static isValidDate(date: string): any {
    const age = this.getAge(date);
    if (age < 13) throw Error("You must be 13 years old or more");
    return new Date(date).toISOString();
  }
}
