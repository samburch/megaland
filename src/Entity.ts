import { v4 as uuidv4 } from "uuid";

export abstract class Entity {
  protected readonly id: string;
  constructor() {
    this.id = uuidv4();
  }
}
