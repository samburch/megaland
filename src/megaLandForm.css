:root {
  font-family: Arial, Helvetica, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;

}

body {
  margin: 0;
  display: flex;
  place-items: center;
  min-width: 320px;
  min-height: 100vh;
}

.form-container {
  margin: 0 auto;
  padding: 2em;
  background-color: #f8f8f8;
  border-radius: 16px;
  max-width: 640px;
  box-shadow: 0 10px 20px -8px rgba(0, 0, 0,.5);
  transition: 0.2s;
}

.form-heading {
  text-align: center;
  width: 80%;
  margin: auto;
}

.form-heading h1 {
  font-size: 2.4em;
  line-height: 1.1;
  margin: 0;
  color: #0084ff;
}


form {
  flex-direction: column;
}

label {
  display: flex;
  margin: .5em 0 0 0;
}

input {
  display: flex;
  width: 100%;
  box-sizing: border-box;
  border-style: solid;
  border-width: 1px;
  border-color: #ddd;
  border-radius: 5px;
  padding: 0.5rem 0.3rem;
  font-family: Arial, Helvetica, sans-serif;
  color:#242424;
}

.input-field > small {
  color: red;
}

.input-field > .success {
  border: 2px solid green;
}

.input-field > .error {
  border: 2px solid red;
}

#firstname, #lastname {
  text-transform: capitalize;
}

#firstname::placeholder, #lastname::placeholder  {
  text-transform: none;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  margin: 1em 0 0 0;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #ffe100;
  cursor: pointer;
  transition: border-color 0.25s;
  width: 100%;
}
button:hover {
  background-color: #ffee00;
  transition: 0.5s;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
  background-color: #fff;
}

input:focus, select:focus{
  outline: 4px auto -webkit-focus-ring-color;
  border-color: #535bf2;
}
