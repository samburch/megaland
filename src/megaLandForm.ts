export interface User {
  firstName: string;
  lastName: string;
  dob: Date;
  email: string;
  password: string;
}

export class UserRegistration implements User {
  constructor(
    public firstName: string,
    public lastName: string,
    public dob: Date,
    public email: string,
    public password: string
  ) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.dob = dob;
    this.email = email;
    this.password = password;
  }

  get age(): number {
    return UserRegistration.calculateAge(this.dob);
  }

  static validateInput(
    input: HTMLInputElement,
    message: string,
    valid: boolean
  ): boolean {
    const error = <HTMLDivElement>input.parentNode!.querySelector("small");
    error.innerText = message;
    input.className = valid ? "success" : "error";
    return !valid;
  }

  static formatName(value: string): string {
    if (value[0] === undefined) return "Anonymous";
    return value[0].toUpperCase() + value.slice(1);
  }

  static calculateAge(value: Date): number {
    const todaysDate = new Date().getFullYear();
    const selectedDate = new Date(value).getFullYear();
    const age: number = todaysDate - selectedDate;
    return age;
  }

  static validate(input: HTMLInputElement): boolean {
    const { type, value }: { type: string; value: string } = input;
    switch (type) {
      case "text":
        if (input.id === "firstname" && value.trim() === "") {
          this.validateInput(input, "You must enter your first name", false);
          return false;
        } else if (input.id === "lastname" && value.trim() === "") {
          this.validateInput(input, "You must enter your last name", false);
          return false;
        } else {
          this.validateInput(input, "", true);
          this.formatName(value);
          return true;
        }
      case "email":
        const email: string = value.toLowerCase().trim();
        const validEmail: RegExp =
          /^([a-zA-Z0-9_\-\.]+)@([a-zA-Z0-9_\-\.]+)\.([a-zA-Z]{2,5})$/;
        if (email === "") {
          this.validateInput(input, "Email must not be blank", false);
          return false;
        }
        if (!email.match(validEmail)) {
          this.validateInput(
            input,
            "Please enter a valid email address",
            false
          );
          return false;
        } else {
          this.validateInput(input, "", true);
          return true;
        }

      case "date":
        const age = this.calculateAge(new Date(value));
        if (value === "") {
          this.validateInput(input, "You must enter your birthday", false);
          return false;
        }
        if (Math.sign(age) === -1) {
          this.validateInput(
            input,
            "Your birthday must be a past date!",
            false
          );
          return false;
        } else if (age < 13) {
          this.validateInput(
            input,
            "You must be more than 13 years old",
            false
          );
          return false;
        } else {
          this.validateInput(input, "", true);
          return true;
        }

      case "password":
        if (value.length < 5 || value.trim() === "") {
          this.validateInput(input, "Minimum 5 characters", false);
          return false;
        } else {
          this.validateInput(input, "", true);
          return true;
        }
    }
    return false;
  }
}
