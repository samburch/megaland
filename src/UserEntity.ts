import { Entity } from "./Entity";
import { DateOfBirth } from "./DateOfBirth";

interface User {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}

export class UserEntity extends Entity implements User {
  constructor(
    public firstName: string,
    public lastName: string,
    public email: string,
    readonly password: string,
    readonly dateOfBirth: DateOfBirth
  ) {
    super();
    this.firstName = this.createName(firstName);
    this.lastName = this.createName(lastName);
    this.password = this.createPassword(password);
    this.email = this.createEmail(email);
    this.dateOfBirth = this.createDate(<any>dateOfBirth);
  }

  // Getters
  get age() {
    return this.dateOfBirth["age"];
  }

  get userFirstName(): string {
    return this.firstName;
  }

  get userLastName(): string {
    return this.lastName;
  }

  get userEmail(): string {
    return this.email;
  }

  // Validator methods
  static isValidInput(value: string): string {
    if (value.trim() === "") throw Error("Please enter");
    if (value.length > 32) throw Error("Must be less than 32 characters");
    return value;
  }

  static isValidEmail(value: string): boolean {
    if (value.length > 254) throw Error("Email must not exceed 254 characters");
    const email = value.toLowerCase();
    const validEmail: RegExp =
      /^([a-zA-Z0-9_\-\.]+)@([a-zA-Z0-9_\-\.]+)\.([a-zA-Z]{2,5})$/;
    validEmail.test(email);
    return validEmail.test(email);
  }

  static isValidPassword(value: string): string {
    const MIN_LENGTH = 8;
    const MAX_LENGTH = 32;
    const PW_VALIDATION: RegExp =
      /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,}$/;
    if (value === undefined || value === "")
      throw Error("Password must not be empty");
    if (value.length < MIN_LENGTH)
      throw Error(`Password must be minimum of ${MIN_LENGTH}`);
    if (value.length > MAX_LENGTH)
      throw Error(`Password must not exceed ${MAX_LENGTH}`);
    if (!PW_VALIDATION)
      throw Error(
        `Password must be btween ${MIN_LENGTH} - ${MAX_LENGTH} and contain at last 1 uppercase character, 1 digit and 1 symbol  i.e. ( ! % ^ " )`
      );
    return value;
  }

  // Factory functions
  private createDate(value: string): DateOfBirth {
    return new DateOfBirth(value);
  }

  private createEmail(email: string): string {
    return UserEntity.isValidEmail(email) ? email : "Invalid email";
  }

  private createName(value: string): string {
    const name = value[0].toUpperCase() + value.slice(1);
    return UserEntity.isValidInput(name) ? name : "Invalid name";
  }

  private createPassword(value: string): string {
    const password = value;
    return UserEntity.isValidInput(password) ? password : "Invalid password";
  }

  // UI validation
  static validateInput(
    input: HTMLInputElement,
    message: string,
    valid: boolean
  ): boolean {
    const error = <HTMLDivElement>input.parentNode!.querySelector("small");
    error.innerText = message;
    input.className = valid ? "success" : "error";
    return !valid;
  }

  static validate(input: HTMLInputElement): boolean {
    const { type, value }: { type: string; value: string } = input;
    switch (type) {
      case "text":
        if (input.id === "firstname" && value.trim() === "") {
          this.validateInput(input, "You must enter your first name", false);
          return false;
        } else if (input.id === "lastname" && value.trim() === "") {
          this.validateInput(input, "You must enter your last name", false);
          return false;
        } else {
          this.validateInput(input, "", true);
          return true;
        }
      case "email":
        const email: string = value.toLowerCase().trim();
        if (email === "") {
          this.validateInput(input, "Email must not be blank", false);
          return false;
        }
        if (!this.isValidEmail) {
          this.validateInput(
            input,
            "Please enter a valid email address",
            false
          );
          return false;
        } else {
          this.validateInput(input, "", true);
          return true;
        }

      case "date":
        if (value === "") {
          this.validateInput(input, "You must enter your birthday", false);
          return false;
        } else if (0) {
          //FIX THIS
          this.validateInput(
            input,
            "You must be more than 13 years old",
            false
          );
          return false;
        } else {
          this.validateInput(input, "", true);
          return true;
        }

      case "password":
        if (value.length < 5 || value.trim() === "") {
          this.validateInput(input, "Minimum 5 characters", false);
          return false;
        } else {
          this.validateInput(input, "", true);
          return true;
        }
    }
    return false;
  }
}
