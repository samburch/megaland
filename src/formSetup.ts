import "./megaLandForm.css";
import { UserRegistration } from "./megaLandForm";

const registrationForm = document.querySelector<HTMLFormElement>(
  "#megaland-registration"
)!;

registrationForm.addEventListener("change", function (event) {
  const input = event.target as HTMLInputElement;
  UserRegistration.validate(input);
});

registrationForm.addEventListener("submit", function (event) {
  event.preventDefault();

  const formElements = registrationForm.elements as HTMLFormControlsCollection;

  const firstname = UserRegistration.validate(
    formElements[<any>"firstname"] as HTMLInputElement
  );
  const lastname: boolean = UserRegistration.validate(
    formElements[<any>"lastname"] as HTMLInputElement
  );
  const dateOfBirth: boolean = UserRegistration.validate(
    formElements[<any>"dob"] as HTMLInputElement
  );
  const email: boolean = UserRegistration.validate(
    formElements[<any>"email"] as HTMLInputElement
  );
  const password: boolean = UserRegistration.validate(
    formElements[<any>"password"] as HTMLInputElement
  );
  // Make the object properties / values immutable
  if (firstname && lastname && email && dateOfBirth && password) {
    const fn: string = UserRegistration.formatName(
      (formElements[<any>"firstname"] as HTMLInputElement).value
    );
    const ln: string = UserRegistration.formatName(
      (formElements[<any>"lastname"] as HTMLInputElement).value
    );
    const dob: Date = (formElements[<any>"dob"] as HTMLInputElement)
      .value as unknown as Date;
    const em = (formElements[<any>"email"] as HTMLInputElement).value;
    const pw = (formElements[<any>"password"] as HTMLInputElement).value;
    const user = Object.freeze(new UserRegistration(fn, ln, dob, em, pw));
    // Log user
    console.log(user);
    // Get user age
    console.log(user.age);
  }
});
