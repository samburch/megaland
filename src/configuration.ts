import "./megaLandForm.css";
import { UserEntity } from "./UserEntity";

const registrationForm = document.querySelector<HTMLFormElement>(
  "#megaland-registration"
)!;

registrationForm.addEventListener("change", function (event) {
  const input = event.target as HTMLInputElement;
  UserEntity.validate(input);
});

registrationForm.addEventListener("submit", function (event) {
  event.preventDefault();

  const formElements = registrationForm.elements as HTMLFormControlsCollection;

  const firstname = UserEntity.validate(
    formElements[<any>"firstname"] as HTMLInputElement
  );
  const lastname: boolean = UserEntity.validate(
    formElements[<any>"lastname"] as HTMLInputElement
  );
  const dateOfBirth: boolean = UserEntity.validate(
    formElements[<any>"dob"] as HTMLInputElement
  );
  const email: boolean = UserEntity.validate(
    formElements[<any>"email"] as HTMLInputElement
  );
  const password: boolean = UserEntity.validate(
    formElements[<any>"password"] as HTMLInputElement
  );

  if (firstname && lastname && email && dateOfBirth && password) {
    const fn: string = (formElements[<any>"firstname"] as HTMLInputElement)
      .value;
    const ln: string = (formElements[<any>"lastname"] as HTMLInputElement)
      .value;
    const dob = <string>(formElements[<any>"dob"] as HTMLInputElement).value;
    const em = (formElements[<any>"email"] as HTMLInputElement).value;
    const pw = (formElements[<any>"password"] as HTMLInputElement).value;
    const userRegistration = new UserEntity(fn, ln, em, pw, dob);
    console.log(userRegistration);
  }
});
