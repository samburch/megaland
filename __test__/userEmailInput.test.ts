import { describe, expect, test } from "@jest/globals";
import { UserRegistration } from "../src/megaLandForm";

describe("Given my email is valid, I should be able to subscribe", () => {
  test("Returns true if users email is valid", () => {
    document.body.innerHTML = `
        <div class="input-field">
            <label for="email">Email address:</label>
            <input type="email" id="email" autocomplete="email" maxlength="254" placeholder="Enter your email address" value="sam.burch85@gmail.com"/>
            <small></small>            
        </div>
  `;
    document.getElementById("email");
    const input = document.getElementById("email") as HTMLInputElement;
    expect(UserRegistration.validate(input)).toBeTruthy();
  });
  test("Adds success class to email input", () => {
    document.body.innerHTML = `
    <div class="input-field">
        <label for="email">Email address:</label>
        <input type="email" id="email" autocomplete="email" maxlength="254" placeholder="Enter your email address" value="sam.burch85@gmail.com"/>
        <small></small>            
    </div>
  `;
    document.getElementById("email");
    const input = document.getElementById("email") as HTMLInputElement;
    UserRegistration.validate(input);
    expect(input).toMatchSnapshot();
  });
});

describe("Given my email is not valid, I should be able to subscribe", () => {
  test("Returns false if users email not valid", () => {
    document.body.innerHTML = `
          <div class="input-field">
              <label for="email">Email address:</label>
              <input type="email" id="email" autocomplete="email" maxlength="254" placeholder="Enter your email address" value="not a valid email"/>
              <small></small>            
          </div>
    `;
    document.getElementById("email");
    const input = document.getElementById("email") as HTMLInputElement;
    expect(UserRegistration.validate(input)).toBeFalsy();
  });
  test("Apply 'Please enter a valid email address' on input field group", () => {
    document.body.innerHTML = `
    <div class="input-field">
        <label for="email">Email address:</label>
        <input type="email" id="email" autocomplete="email" maxlength="254" placeholder="Enter your email address" value="not a valid email"/>
        <small></small>            
    </div>
  `;
    document.getElementById("email");
    const input = document.getElementById("email") as HTMLInputElement;
    const small = document.querySelector("small");
    UserRegistration.validate(input);
    expect(small?.innerText).toContain("Please enter a valid email address");
  });
  test("Returns false if users email is empty", () => {
    document.body.innerHTML = `
          <div class="input-field">
              <label for="email">Email address:</label>
              <input type="email" id="email" autocomplete="email" maxlength="254" placeholder="Enter your email address" value=""/>
              <small></small>            
          </div>
    `;
    document.getElementById("email");
    const input = document.getElementById("email") as HTMLInputElement;
    expect(UserRegistration.validate(input)).toBeFalsy();
  });
  test("Adds error class to date input", () => {
    document.body.innerHTML = `
    <div class="input-field">
        <label for="email">Email address:</label>
        <input type="email" id="email" autocomplete="email" maxlength="254" placeholder="Enter your email address" value=""/>
        <small></small>            
    </div>
  `;
    document.getElementById("email");
    const input = document.getElementById("email") as HTMLInputElement;
    UserRegistration.validate(input);
    expect(input).toMatchSnapshot();
  });
  test("Apply 'Email must not be blank message' on input field group", () => {
    document.body.innerHTML = `
    <div class="input-field">
        <label for="email">Email address:</label>
        <input type="email" id="email" autocomplete="email" maxlength="254" placeholder="Enter your email address" value=""/>
        <small></small>            
    </div>
  `;
    document.getElementById("email");
    const input = document.getElementById("email") as HTMLInputElement;
    const small = document.querySelector("small");
    UserRegistration.validate(input);
    expect(small?.innerText).toContain("Email must not be blank");
  });
});
