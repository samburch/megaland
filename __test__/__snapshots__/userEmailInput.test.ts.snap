// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given my email is not valid, I should be able to subscribe Adds error class to date input 1`] = `
<input
  autocomplete="email"
  class="error"
  id="email"
  maxlength="254"
  placeholder="Enter your email address"
  type="email"
  value=""
/>
`;

exports[`Given my email is valid, I should be able to subscribe Adds success class to email input 1`] = `
<input
  autocomplete="email"
  class="success"
  id="email"
  maxlength="254"
  placeholder="Enter your email address"
  type="email"
  value="sam.burch85@gmail.com"
/>
`;
